### PR TITLE
Adding X509ZNodeGroupAclProvider support in NettyServerCnxnFactory.

### PR DESCRIPTION
**Pull request by rahulrane50**
Currently NettyServerCnxnFactory only support X509AuthenticationProvider as auth provider but new X509ZNodeGroupAclProvider provides custom authentication and authorization logic. Adding support in NettyServerCnxnFactory so that both auth provider could be provided.

Testing : 
Verified that client is able securely connect to zk server with both auth providers (X509ZNodeGroupAclProvider and X509ZNodeGroupAclProvider)
Verified that client is able to connect to plaintextport after unsetting CLIENT_JVMFLAGS

### CODE REVIEW

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/ServerAuthenticationProvider.java [hunk 1]
```diff
@@ -132,8 +132,7 @@ public int getPerm() {
      */
     public abstract boolean matches(ServerObjs serverObjs, MatchValues matchValues);
 
-    @Override
-    public final KeeperException.Code handleAuthentication(ServerCnxn cnxn, byte[] authData) {
+    public KeeperException.Code handleAuthentication(ServerCnxn cnxn, byte[] authData) {
```
Comment (mgao0) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/ServerAuthenticationProvider.java]: Is there a reason this class needs to be changed?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java [hunk 2]
```diff
@@ -72,14 +72,15 @@ public ZkClientUriDomainMappingHelper(ZooKeeperServer zks) {
 
     this.rootPath =
         X509AuthenticationConfig.getInstance().getZnodeGroupAclClientUriDomainMappingRootPath();
+    LOG.debug("Client URI domain mapping root path: {}", this.rootPath);
```
Comment (mgao0) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java]: nit: To make the logs more consistent, please follow the styles of the loggings originally in this class - Have the class name at the beginning of the log message.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java [hunk 3]
```diff
@@ -125,8 +126,10 @@ private void parseZNodeMapping() {
         try {
           List<String> clientUris =
               zks.getZKDatabase().getChildren(rootPath + "/" + domainName, null, null);
-          clientUris.forEach(
-              clientUri -> newClientUriToDomainNames.computeIfAbsent(clientUri, k -> new HashSet<>()).add(domainName));
+          for (String clientUri : clientUris) {
```
Comment (mgao0) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java]: nit: You can minimize unnecessary changes by doing clientUris.forEach(clientUri -> {...})

**Review by mgao0 — COMMENTED** [review]
Congratulations on your first PR! 
A few minor comments.
Great work on identifying the problem, proposing a solution, and fixing the problem!

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/ServerAuthenticationProvider.java (hunk 1 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/ServerAuthenticationProvider.java]: So handleAuthentication method can not be defined as "final" here because we wanted to "override" it's implementation in "X509ZNodeGroupAclProvider"

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxnFactory.java [hunk 4]
```diff
@@ -423,15 +425,26 @@ public void operationComplete(Future<Channel> future) {
 
                     String authProviderProp = System.getProperty(x509Util.getSslAuthProviderProperty(), "x509");
 
-                    X509AuthenticationProvider authProvider = (X509AuthenticationProvider) ProviderRegistry.getProvider(authProviderProp);
+                    X509AuthenticationProvider authProvider = null;
+                    ServerAuthenticationProvider serverAuthProvider = null;
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxnFactory.java:432]: Could we add a descriptive comment why this is necessary here? I think we should give enough context so anyone who sees this can understand the reason behind this change :)

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxnFactory.java [hunk 5]
```diff
@@ -423,15 +425,29 @@ public void operationComplete(Future<Channel> future) {
 
                     String authProviderProp = System.getProperty(x509Util.getSslAuthProviderProperty(), "x509");
 
-                    X509AuthenticationProvider authProvider = (X509AuthenticationProvider) ProviderRegistry.getProvider(authProviderProp);
+                    // All customer variatoins of AuthenticationProvider should be supported here. Currently
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxnFactory.java]: Fix typo - also a better way to say this here is "All implementations of the AuthenticationProvider interface should be supported here"

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxnFactory.java [hunk 6]
```diff
@@ -423,15 +425,29 @@ public void operationComplete(Future<Channel> future) {
...
+                    // any variation of X509AuthenticationProvider or ServerAuthenticationProvider is supported with
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxnFactory.java]: variation -> custom implementation

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java [hunk 7]
```diff
@@ -125,8 +126,10 @@ private void parseZNodeMapping() {
         try {
           List<String> clientUris =
               zks.getZKDatabase().getChildren(rootPath + "/" + domainName, null, null);
-          clientUris.forEach(
-              clientUri -> newClientUriToDomainNames.computeIfAbsent(clientUri, k -> new HashSet<>()).add(domainName));
+          clientUris.forEach(clientUri -> {
+              LOG.debug("ZkClientUriDomainMappingHelper::parseZNodeMapping(): Adding client uri mapping:{} {}", clientUri, domainName);
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java]: Let's make clear which is which?

`(..."domainName: {}, clientUri: {}", domainName, clientUri);`

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java [hunk 8]
```diff
@@ -72,14 +72,15 @@ public ZkClientUriDomainMappingHelper(ZooKeeperServer zks) {
 
     this.rootPath =
         X509AuthenticationConfig.getInstance().getZnodeGroupAclClientUriDomainMappingRootPath();
+    LOG.debug("ZkClientUriDomainMappingHelper::ClientUriDomainMapping Client URI domain mapping root path: {}", this.rootPath);
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java]: Should we make these all `INFO` level logs? Would this be useful in production?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/ServerAuthenticationProvider.java (hunk 1 above)
Comment (narendly) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/ServerAuthenticationProvider.java]: I still don't understand why this change is required.

The following is my understanding:
`X509ZNodeGroupAclProvider` shouldn't implement `handleAuthentication(ServerCnxn cnxn, byte[] authData)`
It only implements `handleAuthentication(ServerObjs serverObjs, byte[] authData)`.

> we wanted to "override" it's implementation in "X509ZNodeGroupAclProvider"

Can you help me understand who's "we" and where is this implementation overridden in `X509ZNodeGroupAclProvider`?

Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/ServerAuthenticationProvider.java]: Yes i think i misunderstood between both handleAuthentication methods. While trying out things i though it was required but it's not required and hence i reverted it. Thanks for pointing it out :)

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java (hunk 8 above)
Comment (rahulrane50) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java]: Sure i feel it's safe to assume that there would in order of 10s applications per domain name and it would be onetime parsing hence should not bloat logs.

**Review by narendly — APPROVED** [review]
Good work!
